### PR TITLE
Simplify scalers, move to `gluonts.torch.scaler`

### PR DESCRIPTION
**Pull request by lostella**
*Description of changes:* There's no need for scalers to be `torch.nn.Module` since they don't really hold parameters. Also fixes the default `keepdim` of `MeanScaler` for consistency.

cc @kashif 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Please tag this pr with at least one of these labels to make our release process faster:** BREAKING, new feature, bug fix, other change, dev setup

### CODE REVIEW

**Comment by kashif** [comment]
very cool! LGTM!

File: src/gluonts/torch/scaler.py [hunk 1]
```diff
@@ -37,21 +36,12 @@ class MeanScaler(nn.Module):
         minimum possible scale that is used for any item.
     """
 
-    @validated()
```
Comment (jaheba) [inline src/gluonts/torch/scaler.py]: We loose the validation property?

File: src/gluonts/torch/scaler.py [hunk 2]
```diff
@@ -37,21 +36,12 @@ class MeanScaler(nn.Module):
...
-        self.dim = dim
-        self.keepdim = keepdim
-        self.register_buffer("minimum_scale", torch.tensor(minimum_scale))
-        self.register_buffer("default_scale", torch.tensor(default_scale))
-
-    def forward(
+    dim: int = -1
+    keepdim: bool = True
+    default_scale: float = 0.0
+    minimum_scale: float = 1e-10
```
Comment (jaheba) [inline src/gluonts/torch/scaler.py]: I think these need to be `torch.tensor`

File: src/gluonts/torch/scaler.py [hunk 3]
```diff
@@ -141,20 +130,11 @@ class StdScaler(nn.Module):
...
-        assert dim > 0, (
-            "Cannot compute scale along dim = 0 (batch dimension), please"
-            " provide dim > 0"
-        )
-        self.dim = dim
-        self.keepdim = keepdim
-        self.register_buffer("minimum_scale", torch.tensor(minimum_scale))
+    dim: int = -1
+    keepdim: bool = False
+    minimum_scale: float = 1e-5
```
Comment (shubhamkapoor) [inline src/gluonts/torch/scaler.py]: Also, `minimum_scale` should be a `torch.Tensor`.

File: src/gluonts/torch/scaler.py (hunk 1 above)
Comment (lostella) [reply src/gluonts/torch/scaler.py]: Should I use our own `dataclass`?

File: src/gluonts/torch/scaler.py (hunk 3 above)
Comment (lostella) [reply src/gluonts/torch/scaler.py]: Why? I think we can add a `torch.Tensor` to a number

File: src/gluonts/torch/scaler.py (hunk 2 above)
Comment (lostella) [reply src/gluonts/torch/scaler.py]: `torch.clamp` also deals with numbers; for the `default_scale`, isn't this some kind of "immutable" (mind the quotes) property of the object, so that replacing `torch.where` with `if` is not really harmful?

**Comment by lostella** [comment]
~~It's funny that all tests pass on macOS~~

File: src/gluonts/torch/scaler.py (hunk 1 above)
Comment (jaheba) [reply src/gluonts/torch/scaler.py]: Yes, let's give it a try

File: src/gluonts/torch/scaler.py [hunk 4]
```diff
@@ -60,21 +58,14 @@ def forward(
...
-        batch_sum = ts_sum.sum(dim=0)
-        batch_observations = torch.clamp(num_observed.sum(0), min=1)
-        batch_scale = torch.squeeze(batch_sum / batch_observations)
-
-        default_scale = torch.where(
-            self.default_scale > 0.0,
-            self.default_scale,
-            batch_scale,
-        )
+        if self.default_scale is not None:
```
Comment (jaheba) [inline src/gluonts/torch/scaler.py]: What is the effect if this change on tracing?

Comment (lostella) [reply src/gluonts/torch/scaler.py]: I think that, as long as `self.default_scale` stays constant (which it should) then tracing should be fine with it and produce the right code for the model. But I need to verify

Comment (lostella) [reply src/gluonts/torch/scaler.py]: Also according to the warning here https://pytorch.org/docs/stable/generated/torch.jit.trace.html#torch-jit-trace it should be fine as long as the control flow is not affected by the value in input tensors (and I would add: if it’s not otherwise changed throughout the model execution, like the `forward` changing the value of `self.default_scale` here for some reason)

File: src/gluonts/torch/scaler.py (hunk 1 above)
Comment (lostella) [reply src/gluonts/torch/scaler.py]: I'll just keep them validated for now

**Review by jaheba — APPROVED** [review]

**Comment by abdulfatir** [comment]
Sorry for the late comment but any takers for Scalers being of `torch.distributions.transforms.Transform` type, `AffineTransform` in particular. This provides us with the inverse function that you can just apply back to your scaled output and also provides `log_abs_det_jacobian`, if needed for loss computation.

**Comment by kashif** [comment]
@abdulfatir on the output side the distribution is an AffineTransform which takes the loc and scale from the scalers... but you mean more general scalers? And I believe currently it's an appropriate place to have the scalers (in the model) since then the model can use the loc and scale as input as well (instead of just on the emission side).

**Comment by abdulfatir** [comment]
Currently, I don't have an example of a general scaler in mind, but yes, using `torch.distributions.transforms.Transform` can provide that flexibility. Regarding access of scaler params (loc, scale) in the models, they can also be accessed from the scaler/transform object's properties. 

The primary benefit IMO is clarity. You have a scaler (Transform) that normalizes the data and then use its inverse (and `log_abs_det_jacobian`) at the output side. 
@kashif

**Comment by abdulfatir** [comment]
Also, inside models I think we should be able to provide a `Scaler()` object instead of a boolean `scaling` as currently implemented.

**Comment by lostella** [comment]
@abdulfatir I think it makes sense to consider this. The (log) scale should be accessible via `log_abs_det_jacobian` if one needs that as input to the model, but not the location. I’ll give it a try and share how that goes

**Comment by abdulfatir** [comment]
@lostella If the transform is of `AffineTransform` type, then we can access the `.loc` and `.scale` attributes. How a model handles/uses transforms could be their internal matter. For instance, for DeepAR like models we can restrict the scale to be of `AffineTransform` type and freely access these properties.

**Comment by abdulfatir** [comment]
For completeness, I would like to add that the inverse transform (which we would need at the output side) is available via the `.inv` attribute.

```python
import torch
from torch.distributions.transforms import AffineTransform

tr = AffineTransform(10., 1.)
inv_tr = tr.inv

x = torch.rand(2, 3, 4)
y = tr(x)

torch.allclose(x, inv_tr(y), atol=1e-5) # True
```

**Comment by lostella** [comment]
> @lostella If the transform is of `AffineTransform` type, then we can access the `.loc` and `.scale` attributes. How a model handles/uses transforms could be their internal matter. For instance, for DeepAR like models we can restrict the scale to be of `AffineTransform` type and freely access these properties.

The catch is that these scaling operations are not really affine transformations of the data: for some array `x`, `x` and `2*x` get scaled down to the same vector. They are once you fix the `loc` and `scale`, so you can access them as properties of the transformation, but `scale` (and possibly `loc`) depend on the input data here.

**Comment by lostella** [comment]
> Also, inside models I think we should be able to provide a Scaler() object instead of a boolean scaling as currently implemented.

@abdulfatir agreed! We can do that in a separate PR
